### ISSUE TITLE
[CRASH] - FBLPromise.m causing several crashes

### ISSUE THREAD

**Issue by JMCPH** [firebase/firebase-ios-sdk#5998]
### Step 1: Describe your environment

  * Xcode version: 11.5
  * Firebase SDK version: 6.27.0
  * Firebase Component: Core, Firestore, Messaging
  * Component version: 6.27.0
  * Installation method: `CocoaPods`

#### Steps to reproduce:

We are experiencing two crashes regarding the `FBLPromise` class.

<img width="1263" alt="Screenshot 2020-07-08 at 10 57 48" src="https://user-images.githubusercontent.com/5514628/86899117-dfb0e500-c109-11ea-836f-6e33e547355a.png">

We have not been able to replicate the crash, but many of our users are experiencing this bug. 
I have attached the stack trace of the two issues below.

What is common for most of our users is that they are experiencing the crash, right after buying our in-app purchase.
<img width="1234" alt="Screenshot 2020-07-08 at 11 02 34" src="https://user-images.githubusercontent.com/5514628/86899676-9e6d0500-c10a-11ea-8dde-a99a1fbbc1e6.png">

#### Relevant Code:
The stack trace of the first crash:
```
Crashed: com.google.perf.FPREventsQueue
0  libsystem_platform.dylib       0x1a7b7d5e8 _os_unfair_lock_corruption_abort + 36
1  libsystem_platform.dylib       0x1a7b79f50 _os_unfair_lock_lock_slow + 288
2  libobjc.A.dylib                0x1a7bab92c objc_sync_enter + 32
3  FBLPromises                    0x104e5f308 -[FBLPromise observeOnQueue:fulfill:reject:] + 215 (FBLPromise.m:215)
4  FBLPromises                    0x104e5f81c -[FBLPromise chainOnQueue:chainedFulfill:chainedReject:] + 281 (FBLPromise.m:281)
5  FBLPromises                    0x104e5a4a0 -[FBLPromise(RecoverAdditions) onQueue:recover:] + 31 (FBLPromise+Recover.m:31)
6  FBLPromises                    0x104e5a3e8 -[FBLPromise(RecoverAdditions) recover:] + 24 (FBLPromise+Recover.m:24)
7  tiimo                          0x10456fd88 -[FIRInstallationsIDController createGetInstallationItemPromise] + 156 (FIRInstallationsIDController.m:156)
8  tiimo                          0x10456fb48 __113-[FIRInstallationsIDController initWithGoogleAppID:appName:installationsStore:APIService:IIDStore:IIDTokenStore:]_block_invoke + 120 (FIRInstallationsIDController.m:120)
9  tiimo                          0x104574494 -[FIRInstallationsSingleOperationPromiseCache getExistingPendingOrCreateNewPromise] + 49 (FIRInstallationsSingleOperationPromiseCache.m:49)
10 tiimo                          0x10456fc7c -[FIRInstallationsIDController getInstallationItem] + 147 (FIRInstallationsIDController.m:147)
11 tiimo                          0x10456bc54 -[FIRInstallations installationIDWithCompletion:] + 174 (FIRInstallations.m:174)
12 tiimo                          0x1045b0db0 -[FPRClient processAndLogEvent:] + 4336373168
13 tiimo                          0x1045b0a40 __29-[FPRClient logNetworkTrace:]_block_invoke + 4336372288
14 libdispatch.dylib              0x1a7b399a8 _dispatch_call_block_and_release + 24
15 libdispatch.dylib              0x1a7b3a524 _dispatch_client_callout + 16
16 libdispatch.dylib              0x1a7ae69d8 _dispatch_lane_serial_drain$VARIANT$mp + 916
17 libdispatch.dylib              0x1a7ae7294 _dispatch_lane_invoke$VARIANT$mp + 416
18 libdispatch.dylib              0x1a7af078c _dispatch_workloop_worker_thread + 588
19 libsystem_pthread.dylib        0x1a7b8bb74 _pthread_wqthread + 272
20 libsystem_pthread.dylib        0x1a7b8e740 start_wqthread + 8

```

The stacktrace of the second crash:
```
Crashed: com.apple.main-thread
0  libsystem_platform.dylib       0x18c1771d4 _os_unfair_lock_corruption_abort + 36
1  libsystem_platform.dylib       0x18c175084 _os_unfair_lock_lock_slow + 320
2  libobjc.A.dylib                0x18c19dc14 objc_sync_enter + 36
3  FBLPromises                    0x101656a08 -[FBLPromise fulfill:] + 85 (FBLPromise.m:85)
4  FBLPromises                    0x10165799c __56-[FBLPromise chainOnQueue:chainedFulfill:chainedReject:]_block_invoke + 270 (FBLPromise.m:270)
5  FBLPromises                    0x101657a40 __56-[FBLPromise chainOnQueue:chainedFulfill:chainedReject:]_block_invoke.48 + 275 (FBLPromise.m:275)
6  libdispatch.dylib              0x18c112ec4 _dispatch_call_block_and_release + 32
7  libdispatch.dylib              0x18c11433c _dispatch_client_callout + 20
8  libdispatch.dylib              0x18c120688 _dispatch_main_queue_callback_4CF + 968
9  CoreFoundation                 0x18c3ef6b0 __CFRUNLOOP_IS_SERVICING_THE_MAIN_DISPATCH_QUEUE__ + 16
10 CoreFoundation                 0x18c3ea2c8 __CFRunLoopRun + 1708
11 CoreFoundation                 0x18c3e98f4 CFRunLoopRunSpecific + 480
12 GraphicsServices               0x196800604 GSEventRunModal + 164
13 UIKitCore                      0x1905bd358 UIApplicationMain + 1944
14 tiimo                          0x100ba25e0 main + 25 (SignupRouter.swift:25)
15 libdyld.dylib                  0x18c2652dc start + 4
```

**Comment by maksymmalyhin** [comment]
@JMCPH I'm sorry you've encountered the issue. From the first glance the issue looks very similar to #4790 which was related to using `objc_sync_enter` and `objc_sync_exit` in the application Swift (see the [comment](https://github.com/firebase/firebase-ios-sdk/issues/4790#issuecomment-581960788) for more details).

Could you please confirm if `objc_sync_...` methods are used in the code of your app or by another SDK?

**Comment by JMCPH** [comment]
@maksymmalyhin thank you for your quick reply.

I have gone through the project and found that it was used in the Mixpanel-swift SDK we are using.

Is there any interfering with Mixpanel and Firebase?

**Comment by maksymmalyhin** [comment]
Method `objc_sync_...` are not intended to be used directly and are implementation details of Objective-C `@synchronized` directive. Using `objc_sync_...` methods may brake any valid Objective-C code that uses `@synchronized` - this is exactly what you experience. 
I would suggest contacting Mixpanel-swift SDK with the issue details and point out to the unsupported usage of Objective-C runtime API (`objc_sync_enter` and `objc_sync_exit` methods).

**Comment by maksymmalyhin** [comment]
I'm closing the ticket as it doesn't look actionable from Firebase side. Feel free to reopen with more details if needed.

**Comment by JMCPH** [comment]
@maksymmalyhin I have contacted Mixpanel and created a issue here:
https://github.com/mixpanel/mixpanel-swift/issues/391 

They believe it is indeed caused by FirebaseCore, so I'm not sure who is right here.

**Comment by juanjoguevara** [comment]
We have experience the same issue in our app.
```Crashed: com.apple.main-thread
0  libsystem_platform.dylib       0x1b69bf1d4 _os_unfair_lock_corruption_abort + 36
1  libsystem_platform.dylib       0x1b69bd084 _os_unfair_lock_lock_slow + 320
2  libobjc.A.dylib                0x1b69e5c14 objc_sync_enter + 36
3  WeFish                         0x1031e5f48 -[FBLPromise fulfill:] + 85 (FBLPromise.m:85)
4  WeFish                         0x1031e6edc __56-[FBLPromise chainOnQueue:chainedFulfill:chainedReject:]_block_invoke + 268 (FBLPromise.m:268)
5  WeFish                         0x1031e6f80 __56-[FBLPromise chainOnQueue:chainedFulfill:chainedReject:]_block_invoke.50 + 273 (FBLPromise.m:273)
6  libdispatch.dylib              0x1b695aec4 _dispatch_call_block_and_release + 32
7  libdispatch.dylib              0x1b695c33c _dispatch_client_callout + 20
8  libdispatch.dylib              0x1b6968688 _dispatch_main_queue_callback_4CF + 968
9  CoreFoundation                 0x1b6c376b0 __CFRUNLOOP_IS_SERVICING_THE_MAIN_DISPATCH_QUEUE__ + 16
10 CoreFoundation                 0x1b6c322c8 __CFRunLoopRun + 1708
11 CoreFoundation                 0x1b6c318f4 CFRunLoopRunSpecific + 480
12 GraphicsServices               0x1c1048604 GSEventRunModal + 164
13 UIKitCore                      0x1bae05358 UIApplicationMain + 1944
14 WeFish                         0x10279f878 main + 14 (main.m:14)
15 libdyld.dylib                  0x1b6aad2dc start + 4 ```

**Comment by maksymmalyhin** [comment]
**TL;DR** Based on experiments, the crash is not reproducible when methods `objc_sync_enter` and `objc_sync_exit` are not used in the app.

@JMCPH, @juanjoguevara Do you have a chance to test and share your feedback if the crash is reproducible for you when you exclude all `objc_sync_enter` and `objc_sync_exit` invocations from your apps? If you still can reproduce the issue feel free to reopen the ticket attaching a sample project where the crash happens.

**A couple notes here:**
1. `objc_sync_enter` and `objc_sync_exit` methods are not described in Apple API documentation and are **not a part of public API so their behaviour is not guaranteed**. It means that using them is not safe as any other undocumented API.
2. Based on the investigation done for  #4790 the unbalanced `objc_sync_enter` and `objc_sync_exit` calls may break unrelated `@synchronized` statements. I GUESS, it happens due to reuse of a lock in a broken state via the lock cache implemented in [objc-sync.mm](https://opensource.apple.com/source/objc4/objc4-781/runtime/objc-sync.mm.auto.html) which leads to the [lock corruption abort](https://github.com/apple/darwin-libplatform/blob/51ce256579c9b879704c868c9cd906ae54f733b5/src/os/lock.c#L494). I may be mistaken as for the exact reason, but it looks there is a reasonable explanation to such a weird behaviour.
3.  Based on #4790 the crash is not reproducible if the methods `objc_sync_enter` and `objc_sync_exit` are not used in the app.

**Comment by juanjoguevara** [comment]
We are remove MixPanel-Swift from our app and the problem is solved. We now are using the MixPanel Objective C library and the crash stopped.

**Comment by maksymmalyhin** [comment]
@juanjoguevara Thank you for the feedback! I'm happy you found a solution!

**Comment by JMCPH** [comment]
I have talked to Mixpanel and they have fixed the issue in version 2.7.5 of their swift framework.
The problem was that they were using the the objc_sync_enter and objc_sync_exit. They have now changed their approach and is using the GCD instead. I hope this helps.
https://github.com/mixpanel/mixpanel-swift/issues/391
